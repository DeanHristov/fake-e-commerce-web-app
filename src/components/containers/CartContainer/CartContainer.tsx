'use client';

import CartTable from '@/components/containers/CartTable';
import CartTableHeading from '@/components/containers/CartTable/ui/CartTableHeading/CartTableHeading';
import CartTableRow from '@/components/containers/CartTable/ui/CartTableRow/CartTableRow';
import ProductCheckoutPanel from '@/components/containers/ProductCheckoutPanel';
import Widget from '@/components/containers/Widget';
import WishListView from '@/components/containers/WishListView';
import Divider from '@/components/ui/Divider';
import { useSelector } from '@/store';
import { selectShoppingCart, selectWishListSize } from '@/store/slices';
import { ICart, ICartProduct } from '@/types';
import { Utils } from '@/utils/Utils';
import { FC } from 'react';

export interface ICartContainerProps {}

const CartContainer: FC<ICartContainerProps> = ({}) => {
  const shoppingCart: ICart = useSelector(selectShoppingCart);
  const totalWishListItems: number = useSelector(selectWishListSize);
  const products: ICartProduct[] = Object.values(shoppingCart.products);

  return (
    <div className="w-full space-y-4 lg:flex lg:justify-start lg:space-x-3 lg:space-y-0">
      <Widget title="Shopping Cart" className="min-h-[16rem] w-full">
        <CartTable>
          <CartTableHeading />
          {Utils.isEmpty(shoppingCart.products) && (
            <h3 className="px-3 py-3 text-center text-lg text-black mt-16">
              Empty!
            </h3>
          )}

          {products.map((product: ICartProduct) => (
            <CartTableRow
              {...product}
              _id={product._id}
              key={`cart-item-${product._id}`}
              currency={shoppingCart.currency}
            />
          ))}
        </CartTable>
        {totalWishListItems > 0 && (
          <div className="mt-10 space-y-6 ">
            <Divider gradient />
            <WishListView />
          </div>
        )}
      </Widget>
      <ProductCheckoutPanel
        currency={shoppingCart.currency!}
        total={shoppingCart.total}
        discountedTotal={shoppingCart.discountedTotal}
        totalProducts={shoppingCart.totalProducts}
        totalQuantity={shoppingCart.totalQuantity}
      />
    </div>
  );
};
export default CartContainer;
