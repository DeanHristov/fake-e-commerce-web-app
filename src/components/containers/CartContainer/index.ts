import CartContainer, { ICartContainerProps } from './CartContainer';

export type { ICartContainerProps };
export default CartContainer;
