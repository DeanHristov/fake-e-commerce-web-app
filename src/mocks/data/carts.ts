import { ICart } from '@/types';

export const carts: ICart[] = [
  {
    _id: '34e1d41b-384a-4bf4-866e-ab5b268d124c',
    products: {
      '8f12eb53-18c8-46a6-b975-ea0ee8b62d4c': {
        _id: '8f12eb53-18c8-46a6-b975-ea0ee8b62d4c',
        title: 'Spring and summershoes',
        price: 20,
        quantity: 3,
        total: 60,
        discountPercentage: 8.71,
        discountedPrice: 55,
        thumbnail: 'https://i.dummyjson.com/data/products/59/thumbnail.jpg',
        inStock: 5,
      },
    },
    currency: 'EUR',
    total: 2,
    discountedTotal: 1941,
    userId: 'ba500896-80ca-42bd-b7f7-a743b959203e',
    totalProducts: 2,
    totalQuantity: 2,
  },
];
