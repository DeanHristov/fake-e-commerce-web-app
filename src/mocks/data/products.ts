import { IProduct, PRODUCT_CATEGORIES } from '../../types';

const products: IProduct[] = [
  {
    _id: '29a34b48-8343-4e3d-8c21-3548ba249936',
    name: 'Airpods Pro2 Headphones',
    image: '/images/airpods.jpg',
    description:
      "Apple's new H2 chip is the driving force behind AirPods Pro and their improved sound characteristics. It works in unison with a personalized driver and amplifier for crystal clear heights and deep, rich bass with incredible clarity, so every sound is more pronounced than ever",
    brand: 'Apple',
    category: PRODUCT_CATEGORIES.ELECTRONIC,
    price: 399.99,
    countInStock: 10,
    rating: 4.5,
    currency: 'USD',
    reviews: 5,
  },
];

export default products;
