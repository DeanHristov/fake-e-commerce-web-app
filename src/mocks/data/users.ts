import bcrypt from 'bcryptjs';
import { IUser } from '../../types';

const users: IUser[] = [
  {
    name: 'Admin User',
    email: 'admin@example.com',
    password: bcrypt.hashSync('123456', 10),
    isAdmin: true,
  },
  {
    name: 'Regular User',
    email: 'regular-user@example.com',
    password: bcrypt.hashSync('123456', 10),
    isAdmin: false,
  },
];

export default users;
