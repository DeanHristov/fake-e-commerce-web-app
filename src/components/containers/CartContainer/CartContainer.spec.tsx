import { render } from '@testing-library/react';

import CartContainer from './CartContainer';
import { carts } from '@/mocks/data/carts';

describe('Container/CartContainer <CartContainer {...} />', () => {
  it('Should render a default state of the Component', () => {
    const { container } = render(<CartContainer {...carts[0]} />);

    expect(container).toMatchSnapshot();
  });
});
