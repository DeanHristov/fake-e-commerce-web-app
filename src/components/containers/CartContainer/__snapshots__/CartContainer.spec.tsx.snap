// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container/CartContainer <CartContainer {...} /> Should be able to display a product 1`] = `
<div>
  <div
    class="w-full space-y-4 lg:flex lg:justify-start lg:space-x-3 lg:space-y-0"
  >
    <div
      class="bg-white border border-gray-300 rounded-lg p-4 shadow-md inline-block relative min-h-[16rem] w-full"
    >
      <h3
        class="text-base text-black font-bold pb-2"
      >
        Shopping Cart
      </h3>
      <ul
        class="w-full flex border-b border-t border-b-dark border-t-dark"
      >
        <li
          class="w-full px-2 py-2 text-center md:text-left"
        >
          <span
            class=""
          >
            Item
          </span>
        </li>
        <li
          class="w-full max-w-[8rem] px-2 py-2 hidden md:block"
        >
          <span
            class=""
          >
            Price
          </span>
        </li>
        <li
          class="w-full max-w-[8rem] px-2 py-2 hidden md:block"
        >
          <span
            class=""
          >
            Quantity
          </span>
        </li>
        <li
          class="w-full max-w-[12rem] px-2 py-2 hidden md:block"
        >
          <span
            class=""
          >
            Subtotal
          </span>
        </li>
      </ul>
      <ul
        class="w-full h-auto min-h-[6rem] border-b border-b-dark px-2 py-2 md:flex md:justify-start"
      >
        <li
          class="w-full md:flex md:justify-start md:space-x-2 max-h-[10rem] overflow-hidden"
        >
          <img
            alt="Iphone"
            class="w-auto h-auto rounded-lg min-w-[9rem] md:max-w-[8rem] border border-gray-300 mx-auto md:m-0"
            data-nimg="1"
            decoding="async"
            fetchpriority="high"
            height="0"
            sizes="100vw"
            src="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=3840&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F59%2Fthumbnail.jpg&w=3840&q=75 3840w"
            style="color: transparent;"
            width="0"
          />
          <h3
            class="product-heading hover:underline hover:cursor-pointer text-center md:text-left"
          >
            <span
              class="text-black text-sm"
            >
              1x 
            </span>
            Iphone
          </h3>
        </li>
        <li
          class="w-full flex justify-center space-x-2 md:justify-start md:max-w-[8rem]"
        >
          <h3
            class="md:hidden"
          >
            Price: 
          </h3>
          <div
            class=""
          >
            <span
              class="price-label"
            >
              €20.00
            </span>
          </div>
        </li>
        <li
          class="w-2/3 px-2 pt-2 mx-auto md:w-full md:max-w-[8rem] md:p-0"
        >
          <div
            class="flex justify-center items-center bg-gray-100 rounded-md border border-gray-400 undefined"
          >
            <button
              class="text-xl w-20 text-red-400 font-medium border-r border-r-gray-400"
            >
              -
            </button>
            <span
              class="w-full px-1  text-center"
            >
              1
            </span>
            <button
              class="text-xl w-20  text-green-400 font-medium border-l border-l-gray-400"
            >
              +
            </button>
          </div>
        </li>
        <li
          class="w-full md:max-w-[12rem] relative"
        >
          <div
            class="w-full flex items-center justify-center space-x-2 py-2 md:py-0 md:flex-col"
          >
            <h3
              class="md:hidden"
            >
              Subtotal: 
            </h3>
            <div
              class=""
            >
              <span
                class="price-label"
              >
                €20.00
              </span>
            </div>
          </div>
          <div
            class="text-black text-sm text-center md:text-left md:absolute md:bottom-0.5 md:right-0"
          >
            <span
              class="hover:underline hover:cursor-pointer hover:text-dark"
              data-testid="btn-remove"
            >
              Remove
            </span>
             / 
            <span
              class="hover:underline hover:cursor-pointer hover:text-dark"
              data-testid="btn-wish-list"
            >
              Wish list
            </span>
          </div>
        </li>
      </ul>
    </div>
    <div
      class="bg-white border border-gray-300 rounded-lg p-4 shadow-md inline-block relative max-h-[16rem] w-full lg:w-auto"
    >
      <div
        class="p-2"
      >
        <h3
          class="px-3 py-3 text-center text-base text-black border-b border-b-dark"
        >
          1 Items in your Bag
        </h3>
        <p
          class="py-2 mb-4 border-b border-b-dark uppercase flex justify-between items-center"
        >
          <span
            class="font-bold text-md text-black"
          >
            total 
          </span>
          <span
            class="font-medium text-md text-black"
          >
            €1.00
          </span>
        </p>
        <div
          class="relative w-full rounded-md w-full h-10 mb-4 mx-auto max-w-xl border border-gray-500"
        >
          <input
            class="h-full px-6 w-full rounded-lg bg-white text-slate-700 ease-in duration-200 focus:bg-white"
            placeholder="Enter a promocode"
            type="text"
          />
        </div>
        <button
          class="cursor-pointer bg-green-600 text-white font-bold uppercase bg-primary hover:bg-hover hover:shadow-lg transition transform active:scale-90 duration-200 ease-in w-full px-4 py-1.5 flex justify-center items-center gap-2 rounded-none"
        >
          Checkout
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Container/CartContainer <CartContainer {...} /> Should render a default state of the Component 1`] = `
<div>
  <div
    class="w-full space-y-4 lg:flex lg:justify-start lg:space-x-3 lg:space-y-0"
  >
    <div
      class="bg-white border border-gray-300 rounded-lg p-4 shadow-md inline-block relative min-h-[16rem] w-full"
    >
      <h3
        class="text-base text-black font-bold pb-2"
      >
        Shopping Cart
      </h3>
      <ul
        class="w-full flex border-b border-t border-b-dark border-t-dark"
      >
        <li
          class="w-full px-2 py-2 text-center md:text-left"
        >
          <span
            class=""
          >
            Item
          </span>
        </li>
        <li
          class="w-full max-w-[8rem] px-2 py-2 hidden md:block"
        >
          <span
            class=""
          >
            Price
          </span>
        </li>
        <li
          class="w-full max-w-[8rem] px-2 py-2 hidden md:block"
        >
          <span
            class=""
          >
            Quantity
          </span>
        </li>
        <li
          class="w-full max-w-[12rem] px-2 py-2 hidden md:block"
        >
          <span
            class=""
          >
            Subtotal
          </span>
        </li>
      </ul>
      <h3
        class="px-3 py-3 text-center text-lg text-black mt-16"
      >
        Empty!
      </h3>
    </div>
    <div
      class="bg-white border border-gray-300 rounded-lg p-4 shadow-md inline-block relative max-h-[16rem] w-full lg:w-auto"
    >
      <div
        class="p-2"
      >
        <h3
          class="px-3 py-3 text-center text-base text-black border-b border-b-dark"
        >
          0 Items in your Bag
        </h3>
        <p
          class="py-2 mb-4 border-b border-b-dark uppercase flex justify-between items-center"
        >
          <span
            class="font-bold text-md text-black"
          >
            total 
          </span>
          <span
            class="font-medium text-md text-black"
          >
            ---
          </span>
        </p>
        <div
          class="relative w-full rounded-md w-full h-10 mb-4 mx-auto max-w-xl border border-gray-500"
        >
          <input
            class="h-full px-6 w-full rounded-lg bg-white text-slate-700 ease-in duration-200 focus:bg-white"
            disabled=""
            placeholder="Enter a promocode"
            type="text"
          />
        </div>
        <button
          class="cursor-not-allowed opacity-50 active:scale-100 bg-green-600 text-white font-bold uppercase bg-primary hover:bg-hover hover:shadow-lg transition transform active:scale-90 duration-200 ease-in w-full px-4 py-1.5 flex justify-center items-center gap-2 rounded-none"
          disabled=""
        >
          Checkout
        </button>
      </div>
    </div>
  </div>
</div>
`;
